### ISSUE TITLE
SchemaGenerationException when auto schema generation is set to "validate".

### ISSUE THREAD

**Issue by moldziej** [Impetus/kundera#570]
Hello,

There is SchemaGenerationException when:
- schema auto generation is set to "validate", and
- primary key auto generation is used.

To illustrate the problem I've modified cassandra-pelops KunderaQueryTest unit test:

``` diff
diff -ru cassandra-pelops-orig/src/test/java/com/impetus/kundera/client/cassandra/pelops/crud/UserAccount.java cassandra-pelops/src/test/java/com/impetus/kundera/client/cassandra/pelops/crud/UserAccount.java
--- cassandra-pelops-orig/src/test/java/com/impetus/kundera/client/cassandra/pelops/crud/UserAccount.java   2014-03-24 05:37:46.000000000 +0100
+++ cassandra-pelops/src/test/java/com/impetus/kundera/client/cassandra/pelops/crud/UserAccount.java    2014-04-11 11:08:00.973448300 +0200
@@ -24,7 +24,9 @@
 import javax.persistence.FetchType;
 import javax.persistence.OneToMany;
 import javax.persistence.Table;
-
+import javax.persistence.Id;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;

 /**
  * User account entity.
@@ -35,7 +37,10 @@
 @Table(name = "user_account")
 public class UserAccount extends GuidDomainObject
 {
-   
+    @Id
+    @GeneratedValue(strategy = GenerationType.TABLE)
+    private Long id;
+
     @Column(name = "display_name", length = 128)
     private String displayName = null;

diff -ru cassandra-pelops-orig/src/test/resources/META-INF/persistence.xml cassandra-pelops/src/test/resources/META-INF/persistence.xml
--- cassandra-pelops-orig/src/test/resources/META-INF/persistence.xml   2014-03-24 05:37:46.000000000 +0100
+++ cassandra-pelops/src/test/resources/META-INF/persistence.xml    2014-04-11 11:11:00.771426300 +0200
@@ -13,7 +13,7 @@
            <property name="kundera.port" value="9160" />
            <property name="kundera.keyspace" value="UUIDCassandra" />
            <property name="kundera.dialect" value="cassandra" />
-           <property name="kundera.ddl.auto.prepare" value="create" />
+           <property name="kundera.ddl.auto.prepare" value="validate" />
            <property name="kundera.client" value="pelops" />
            <property name="kundera.client.lookup.class"
                value="com.impetus.client.cassandra.pelops.PelopsClientFactory" />

```

I run the test with kundera.ddl.auto.prepare set to create for the first time. When I run next time with "validate" following execption is thrown:

```
Running com.impetus.kundera.query.KunderaQueryTest
log4j:WARN No appenders could be found for logger (org.apache.cassandra.service.CassandraDaemon).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Tests run: 10, Failures: 0, Errors: 10, Skipped: 0, Time elapsed: 8.491 sec <<< FAILURE!
test(com.impetus.kundera.query.KunderaQueryTest)  Time elapsed: 7.818 sec  <<< ERROR!
com.impetus.kundera.configure.schema.SchemaGenerationException: com.impetus.kundera.configure.schema.SchemaGenerationException: Column family kundera_sequences does not exist in keyspace UUIDCassandra
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.validate(CassandraSchemaManager.java:471)
    at com.impetus.kundera.configure.schema.api.AbstractSchemaManager.handleOperations(AbstractSchemaManager.java:262)
    at com.impetus.kundera.configure.schema.api.AbstractSchemaManager.exportSchema(AbstractSchemaManager.java:112)
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.exportSchema(CassandraSchemaManager.java:146)
    at com.impetus.kundera.configure.SchemaConfiguration.configure(SchemaConfiguration.java:187)
    at com.impetus.kundera.configure.ClientMetadataBuilder.buildClientFactoryMetadata(ClientMetadataBuilder.java:48)
    at com.impetus.kundera.persistence.EntityManagerFactoryImpl.configureClientFactories(EntityManagerFactoryImpl.java:404)
    at com.impetus.kundera.persistence.EntityManagerFactoryImpl.configure(EntityManagerFactoryImpl.java:157)
    at com.impetus.kundera.persistence.EntityManagerFactoryImpl.<init>(EntityManagerFactoryImpl.java:131)
    at com.impetus.kundera.KunderaPersistence.createEntityManagerFactory(KunderaPersistence.java:85)
    at javax.persistence.Persistence.createEntityManagerFactory(Unknown Source)
    at javax.persistence.Persistence.createEntityManagerFactory(Unknown Source)
    at com.impetus.kundera.query.KunderaQueryTest.setUp(KunderaQueryTest.java:60)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:44)
(...)
```

There is information that "Column family <b>kundera_sequences</b> does not exist in keyspace UUIDCassandra" but I've checked in cqlsh that kundera_sequences exsists.

Best regards,
Marcin

**Comment by mevivs** [comment]
Looks like column family created via CQL3 whereas schema generation is set for CQL version 2(i.e. thrift). Please set CQL version to 3 and try.

-Vivek

**Comment by moldziej** [comment]
I work on empty Cassandra database.
Keyspace and colum families are created by kundera when I run the test for the first time with kundera.ddl.auto.prepare set to "create".
Then I run the test once again with kundera.ddl.auto.prepare set to "validate".

Do you mean setting CassandraConstants.CQL_VERSION_3_0 on entityManagerFactory ?

**Comment by mevivs** [comment]
Yes. Using Kundera version 2.11?
-Vivek

**Comment by moldziej** [comment]
Yes, 2.11.

Setting CQL3 helps, but could you explain more deeply where is the mismatch? I had to recreate the schema after setting CQL3. To sum up I run multiple times cassandra-pelops KunderaQueryTest, starting from empty Cassandra:
1. kundera.ddl.auto.prepare="create"   - passed
2. kundera.ddl.auto.prepare="validate"  - SchemaGenerationException on <b>kundera_sequences</b>
3. kundera.ddl.auto.prepare="validate"  & CQL3 - SchemaGenerationException on <b>social_profile</b>
4. kundera.ddl.auto.prepare="create"  & CQL3  - passed
5. kundera.ddl.auto.prepare="validate"  & CQL3 - passed

<p>
And one more question. If I have following persistence.xml with CQL3 set on entityManagerFactory 

``` xml
<persistence-unit name="cass_pu">
    <provider>com.impetus.kundera.KunderaPersistence</provider>
    <properties>
        (...)
        <property name="kundera.dialect" value="cassandra" />
        <property name="kundera.ddl.auto.prepare" value="validate" />
        <property name="kundera.client" value="pelops" />
        <property name="kundera.client.lookup.class"
            value="com.impetus.client.cassandra.pelops.PelopsClientFactory" />
    </properties>
```

then what is the transport used?
kundera -> CQL3 -> Pelops lib -> Thrift RPC transport -> Cassandra ?

Thank you,
Marcin.

**Comment by mevivs** [comment]
```
kundera.ddl.auto.prepare="create" - passed
kundera.ddl.auto.prepare="validate" - SchemaGenerationException on kundera_sequences
kundera.ddl.auto.prepare="validate" & CQL3 - SchemaGenerationException on social_profile
kundera.ddl.auto.prepare="create" & CQL3 - passed
kundera.ddl.auto.prepare="validate" & CQL3 - passed
```

Let me verify and get back to you on this.

For ThriftClientFactory and PelopsClientFactory it will be Thrift RPC.But for Cassandra DS driver it will be CQL3 binary protocol. I recommend  to use ThriftClientFactory for thrift rpc and DS driver factory for CQL3 binary protocol.

Hope it helps.

-Vivek

**Comment by moldziej** [comment]
I've checked that in KunderaQueryTest, in 1st and 2nd point described above, CF are <i>not</i> generated with CQL3 but as you can see schema validation fails on kundera_sequences.

**Comment by moldziej** [comment]
Hello again,
I've found out what's the reason of validation failure when primary key autogeneration is used.

When schema in kundera is created, "kundera_sequences" TableInfo is created with type set to "CounterColumnType":

``` java
com.impetus.kundera.configure.SchemaConfiguration.addTableGenerator(ApplicationMetadata, String, List<TableInfo>, EntityMetadata, Class, String)
{
    //(...)
    TableInfo tableGeneratorDiscriptor = new TableInfo(keyValue.getTableDiscriptor().getTable(),
                    "CounterColumnType", String.class, idName);
    //(...)
}
```

Then, when schema is exported to schema manager, it is validated with schema retrieved from Cassandra:

``` java
com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.onValidateTable(KsDef ksDef, TableInfo tableInfo)
{
    boolean tablefound = false;
    for (CfDef cfDef : ksDef.getCf_defs())
    {
        if (cfDef.getName().equals(tableInfo.getTableName())
                && (cfDef.getColumn_type().equals(ColumnFamilyType.getInstanceOf(tableInfo.getType()).name())))
        {

    //(...)

    if (!tablefound)
    {
        throw new SchemaGenerationException("Column family " + tableInfo.getTableName()
                + " does not exist in keyspace " + databaseName + "", "Cassandra", databaseName,
                tableInfo.getTableName());
    }   
}
```

In cfDef (retrieved from Cassandra) column type is <b>"Standard"</b>.
While in tableInfo the type is "CounterColumnType", wchis is converted to <b>"Super"</b> in:

``` java
    ColumnFamilyType.getInstanceOf(tableInfo.getType()).name()
```

The types does not match. That's why table is not found in onValidateTable function and validation fails with SchemaGenerationException.

Could you suggest a proper way to fix this mismatch?

Regards,
Marcin

**Comment by mevivs** [comment]
@moldziej 
You are right. Team is working on to fix it on current trunk.Will soon post an update.

-Vivek

**Comment by chhavigangwal** [comment]
Fixed in Kundera-2.11.1
